src/commands/ext-dev-deprecate.ts

```diff
@@ -0,0 +1,77 @@
+import { ensureExtensionsApiEnabled, logPrefix } from "../extensions/extensionsHelper";
+import { deprecateExtensionVersion, listExtensionVersions } from "../extensions/extensionsApi";
+import * as refs from "../extensions/refs";
+import { parseVersionPredicate, compareVersions } from "../extensions/versionHelper";
+import { requireAuth } from "../requireAuth";
+import { FirebaseError } from "../error";
+import * as utils from "../utils";
+
+/**
+ * Deprecate all extension versions that match the version predicate.
+ */
+export default new Command("ext:dev:deprecate <extensionRef> <versionPredicate>")
+  .description("deprecate extension versions that match the version predicate")
+  .option("-m, --message <deprecationMessage>", "deprecation message")
+  .option(
+    "-f, --force",
+    "override deprecation message for existing deprecated extension versions that match"
+  )
+  .before(requireAuth)
+  .before(ensureExtensionsApiEnabled)
+  .action(async (extensionRef: string, versionPredicate: string, options: any) => {
+    const { publisherId, extensionId, version } = refs.parse(extensionRef);
+    const baseExtensionRef = publisherId + "/" + extensionId;
+    if (version) {
+      throw new FirebaseError(
+        `The input extension reference must be of the format ${clc.bold(
+          "<publisherId>/<extensionId>"
+        )}. Version should be supplied in the version predicate argument.`
+      );
+    }
+    if (!publisherId || !extensionId) {
+      throw new FirebaseError(
+        `Error parsing publisher ID and extension ID from extension reference '${clc.bold(
+          extensionRef
+        )}'. Please use the format '${clc.bold("<publisherId>/<extensionId>")}'.`
+      );
+    }
+    const { comparator, targetSemVer } = parseVersionPredicate(versionPredicate);
+    const extensionVersions = await listExtensionVersions(baseExtensionRef);
+    const filteredExtensionVersions = extensionVersions.filter((extensionVersion) => {
+      const semVer = refs.parse(extensionVersion.ref).version!;
+      if (
+        (extensionVersion.state === "DEPRECATED" && !options.force) ||
+        !compareVersions(semVer, targetSemVer, comparator)
+      ) {
+        return false;
+      }
+      const message =
+        extensionVersion.state === "DEPRECATED" ? ", will overwrite deprecation message" : "";
+      utils.logLabeledBullet(extensionVersion.ref, extensionVersion.state + message);
+      return true;
+    });
+    if (filteredExtensionVersions.length > 0) {
+      const confirmMessage =
+        "You are about to deprecate these extension version(s). Do you wish to continue?";
+      const consent = await promptOnce({
+        type: "confirm",
+        message: confirmMessage,
+        default: false,
+      });
+      if (!consent) {
+        throw new FirebaseError("Deprecation canceled.");
+      }
+    } else {
+      throw new FirebaseError("No extension versions matched the version predicate.");
+    }
+    await utils.allSettled(
```

Comment:
What happens if one of these calls errors? Do we report it correctly with this set up?

Reply:
I believe if one call errors out, the rest will finish successfully but the user will receive an error instead of the success message. Still not clear that some may have succeeded though. Do you think this is a problem?

Reply:
Ok - I think its Ok as is for now, but lets follow up and improve this error handling. Ideally, we should build a little rig to catch all the errors, and report which version were successfully deprecated/undeprecated, and which had errors.

For now, I suspect that the error rate once we get to this point is so low that it is not important to have nice error handling.

Reply:
I agree, not ideal UX right now, but the saving grace is that if let's say two of three versions fail to get deprecated and the user receives a "failed to deprecate" message, simply retrying the previous command will correctly deprecate the remaining versions whether it's with `--force` or not.